lib/lightning/maintenance/partition_table_service.ex

```diff
@@ -0,0 +1,185 @@
+      for {key, val} <- shift_options,
+          into: [],
+          do: {String.to_existing_atom(key), val}
+
+    upper_bound = Timex.shift(DateTime.utc_now(), shift_options)
+
+    remove_empty("work_orders", upper_bound)
+  end
+
+  def add_headroom(:all, weeks) when is_integer(weeks) do
+    add_headroom(:work_orders, weeks) |> log_partition_creation()
+  end
+
+  def add_headroom(:work_orders, weeks) when is_integer(weeks) do
+    proposed_tables = tables_to_add("work_orders", weeks)
+
+    :ok =
+      Enum.each(proposed_tables, fn {partition_name, from, to} ->
+        {
+          Repo.query(create_query(partition_name, "work_orders", from, to))
+        }
+      end)
+
+    proposed_tables
+  end
+
+  def tables_to_add(table, weeks) do
+    today = Date.utc_today()
+
+    existing_tables = get_partitions(table)
+
+    Lightning.AdminTools.generate_iso_weeks(today, today |> Date.add(weeks * 7))
+    |> Enum.map(&to_partition_details(table, &1))
+    |> Enum.reject(fn {name, _from, _to} ->
+      Enum.find(existing_tables, &String.equivalent?(name, &1))
+    end)
+  end
+
+  def get_partitions(parent) do
+    %Postgrex.Result{rows: rows} =
+      Repo.query!(
+        ~S[
+          SELECT CAST(inhrelid::regclass AS text) AS child
+          FROM   pg_catalog.pg_inherits
+          WHERE  inhparent = $1::text::regclass;
+        ],
+        [parent]
+      )
+
+    rows |> List.flatten()
+  end
+
+  @doc """
+  Drops empty partition tables that have an upper partition bound less than the
+  date given.
+
+  This bound is the `TO` part of the partition:
+
+  ```
+  FOR VALUES FROM ('2020-12-28 00:00:00') TO ('2021-01-04 00:00:00')
+  ```
+  """
+  def remove_empty(parent, upper_bound) do
+    parent
+    |> find_range_partitions
+    |> partitions_older_than(upper_bound)
+    |> Enum.each(&drop_partition(parent, &1))
+  end
+
+  def find_range_partitions(parent) do
+    Repo.query!(
+      ~S[
+        SELECT
+          pt.relname AS partition_name,
+          pg_get_expr(pt.relpartbound,
+          pt.oid,
+          TRUE) AS partition_expression
+        FROM
+          pg_class base_tb
+        JOIN pg_inherits i ON
+          i.inhparent = base_tb.oid
+        JOIN pg_class pt ON
+          pt.oid = i.inhrelid
+        WHERE
+          base_tb.oid = $1::text::regclass
+          AND pg_get_expr(
+            pt.relpartbound,
+            pt.oid,
+            TRUE
+          ) != 'DEFAULT'
+      ],
+      [parent]
+    ).rows
+  end
+
+  def partitions_older_than(partitions, bound) do
+    partitions
+    |> Enum.map(fn [table, range_expression] ->
+      [_, to_as_string] =
+        ~r/TO \('(.+)'\)/
+        |> Regex.run(range_expression)
+
+      [table, NaiveDateTime.from_iso8601!(to_as_string)]
+    end)
+    |> Enum.filter(fn [_table, to] -> to < bound end)
+    |> Enum.map(fn [table, _to] -> table end)
+  end
+
+  def drop_partition(parent, partition) do
+    unless valid_chars?(parent) && valid_chars?(partition) do
```

Comment:
String interpolation in queries makes me nervous :)- so I added some simple validation against the more obvious forms of injection (as I know that we can't bind params in DDL statements) just in case this method ends up being used someplace else.